src/main/java/de/tum/in/www1/artemis/service/notifications/push_notifications/PushNotificationService.java

```diff
@@ -0,0 +1,175 @@
+import de.tum.in.www1.artemis.service.notifications.InstantNotificationService;
+
+/**
+ * Wraps the sending of iOS and Android Notifications to the Relay Service
+ * Implements the encryption of the payload
+ */
+public abstract class PushNotificationService implements InstantNotificationService {
+
+    private static SecureRandom random = new SecureRandom();
+
+    private static Cipher cipher;
+
+    static {
+        try {
+            cipher = Cipher.getInstance(Constants.PUSH_NOTIFICATION_ENCRYPTION_ALGORITHM);
+        }
+        catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(PushNotificationService.class);
+
+    private static final Gson gson = new Gson();
+
+    private final RestTemplate restTemplate;
+
+    protected PushNotificationService(RestTemplate restTemplate) {
+        this.restTemplate = restTemplate;
+    }
+
+    /**
+     * Send all the notifications requests to the endpoint. Potentially, optimize the sending using a batched request.
+     *
+     * @param requests           the requests previously built using buildSendRequest
+     * @param relayServerBaseUrl the url of the relay
+     */
+    void sendNotificationRequestsToEndpoint(List<RelayNotificationRequest> requests, String relayServerBaseUrl) {
+        var futures = requests.stream()
+                .map(request -> CompletableFuture.runAsync(() -> sendSpecificNotificationRequestsToEndpoint(Collections.singletonList(request), relayServerBaseUrl))).toList()
+                .toArray(new CompletableFuture[0]);
+
+        CompletableFuture.allOf(futures);
+    }
+
+    @Async
+    void sendRelayRequest(String body, String relayServerBaseUrl) {
+        RetryTemplate template = RetryTemplate.builder().exponentialBackoff(1000, 4, 60 * 1000).retryOn(RestClientException.class).maxAttempts(4).build();
+
+        try {
+            template.execute((RetryCallback<Void, RestClientException>) context -> {
+                HttpHeaders httpHeaders = new HttpHeaders();
+                httpHeaders.setContentType(MediaType.APPLICATION_JSON);
+                HttpEntity<String> httpEntity = new HttpEntity<>(body, httpHeaders);
+                restTemplate.postForObject(relayServerBaseUrl + getRelayPath(), httpEntity, String.class);
+
+                return null;
+            });
+        }
+        catch (RestClientException e) {
+            log.error("Could not send " + getDeviceType().toString() + " notifications");
+        }
+    }
+
+    @Override
+    public final void sendNotification(Notification notification, User user, Object notificationSubject) {
+        sendNotification(notification, Collections.singletonList(user), notificationSubject);
+    }
+
+    @Override
+    @Async
+    public void sendNotification(Notification notification, List<User> users, Object notificationSubject) {
```

Comment:
Please check if we need some javadoc here. Couldn't hurt to explain what's happening here with the encryption etc